api-lib/email/unsubscribe.ts

```diff
@@ -0,0 +1,47 @@
+import { createHmac } from 'crypto';
+
+import { HMAC_SECRET } from '../config';
+import { UnauthorizedError } from '../HttpError';
+
+type EmailType = 'product' | 'transactional' | 'notification';
+
+export function isEmailType(emailType: string): emailType is EmailType {
+  return ['product', 'transactional', 'notification'].includes(emailType);
+}
+
+export function genToken(
+  profileId: string,
+  email: string,
+  emailType: string
+): string {
+  const token = genHmac(profileId, email, emailType);
+
+  const params = new URLSearchParams({ profileId, email, emailType, token });
+  return params.toString();
+}
+
+export function decodeToken(encodedString: string): {
+  profileId: string;
+  email: string;
+  emailType: string;
+} {
+  const params = new URLSearchParams(encodedString);
+  const profileId = params.get('profileId');
+  const email = params.get('email');
+  const token = params.get('token');
+  const emailType = params.get('emailType');
+  if (!profileId || !email || !token || !emailType || !isEmailType(emailType)) {
+    throw new UnauthorizedError('Invalid unsubscribe token');
+  }
+
+  const generatedToken = genHmac(profileId, email, emailType);
+  if (token !== generatedToken) {
```

Comment:
use https://nodejs.org/dist/latest-v6.x/docs/api/crypto.html#crypto_crypto_timingsafeequal_a_b not ==

Reply:
@crabsinger  I am currently using timingSafeEqual. I converted the string into a buffer first.
I have tested the functionality after.